### ISSUE TITLE
Robot doesn't move during inference when using validation split during training

### ISSUE THREAD

**Issue by kagyvro48** [huggingface/lerobot#2851]
Hi team,

I'm running into a weird issue that's been blocking me for a while. When I train policies (ACT, Pi0, SmolVLA) with a validation split, the robot completely freezes during inference - it just doesn't move at all. But when I train on the full dataset without validation, everything works perfectly.

## The Problem
I have 301 episodes of teleoperation data (15 seconds each) for a weed-pulling task. Here's what happens:
Scenario 1 - With validation (BROKEN):
- Split data: 85% training (256 episodes) + 15% validation (45 episodes)
- Training runs fine, validation loss looks reasonable on WandB
- But at inference: robot doesn't move, actions seem to be near zero
Scenario 2 - Without validation (WORKS):
- Use all 301 episodes for training
- Same hyperparameters, same everything
- Robot works perfectly at inference
I've tested this on ACT, Pi0, and SmolVLA - same issue every time.

## What I think is happening
I suspect there's a normalization statistics mismatch. When you specify `val_episodes`, the dataset still loads the normalization stats from `info.json` which were computed on all 301 episodes. But the training loop only sees 256 episodes, and validation sees 45 different episodes. This creates an inconsistency that breaks action prediction during inference.
Previously I trained on 250 episodes without validation and it worked great, so it's definitely something related to the validation split mechanism.

## Why this matters
I’m analyzing different VLA architectures, and I’d like to know if there are any limitations regarding the validation loss at this stage. Right now I'm stuck either:
- Using validation and getting broken models
- Not using validation and having no proper evaluation methodology
I had to override some code in [lerobot_train.py](lerobot/src/lerobot/scripts/lerobot_train.py)  to add validation support (around lines 396-406) since it wasn't fully implemented, so maybe I'm doing something wrong? Or is this a known limitation?
Would really appreciate any guidance here. Is validation officially supported? If not, what's the recommended way to evaluate models during training?

Thanks!

**Comment by Saad-Mallebhari** [comment]
@kagyvro48 
I’ve been seeing the same behavior and wanted to share some observations.

When using `val_episodes`, training and validation run normally, but at inference time the policy outputs collapse (near-zero actions). If I train on the full dataset without a validation split, inference works as expected. I’ve observed this consistently across ACT, Pi0, and SmolVLA.

From what I can tell, this seems related to normalization. The stats in `info.json` are computed over the full dataset, while training only sees the training subset when validation is enabled. That mismatch likely causes the policy to operate in a different scale at inference time.

One side effect of this is that actions which are “reasonable” in the model’s normalized space may get scaled down to values that fall below the robot’s effective actuation threshold (deadband), which would explain the frozen behavior despite non-zero predictions.

Another thing worth checking: since validation requires some manual plumbing right now, it’s possible that the training loop or overrides (e.g. in `lerobot_train.py`) are inadvertently mixing normalization stats or switching the model into a state that affects scaling (e.g. BatchNorm / running stats, if present).

If validation is meant to be supported, it seems like normalization should be recomputed strictly from the training split and then reused consistently for training, validation, and inference. As a quick sanity check, one way to confirm this would be:
- recompute normalization using only the training episodes
- save those stats separately
- run inference using that version

If the policy behaves correctly in that setup, it would strongly suggest the issue is due to global normalization rather than model behavior.
Is validation currently expected to work in this configuration, or is the recommended workflow still to train on the full dataset and evaluate separately? Happy to provide a minimal repro or help test a fix if useful.

**Comment by kagyvro48** [comment]
> [@kagyvro48](https://github.com/kagyvro48) I’ve been seeing the same behavior and wanted to share some observations.
> 
> When using `val_episodes`, training and validation run normally, but at inference time the policy outputs collapse (near-zero actions). If I train on the full dataset without a validation split, inference works as expected. I’ve observed this consistently across ACT, Pi0, and SmolVLA.
> 
> From what I can tell, this seems related to normalization. The stats in `info.json` are computed over the full dataset, while training only sees the training subset when validation is enabled. That mismatch likely causes the policy to operate in a different scale at inference time.
> 
> One side effect of this is that actions which are “reasonable” in the model’s normalized space may get scaled down to values that fall below the robot’s effective actuation threshold (deadband), which would explain the frozen behavior despite non-zero predictions.
> 
> Another thing worth checking: since validation requires some manual plumbing right now, it’s possible that the training loop or overrides (e.g. in `lerobot_train.py`) are inadvertently mixing normalization stats or switching the model into a state that affects scaling (e.g. BatchNorm / running stats, if present).
> 
> If validation is meant to be supported, it seems like normalization should be recomputed strictly from the training split and then reused consistently for training, validation, and inference. As a quick sanity check, one way to confirm this would be:
> 
> * recompute normalization using only the training episodes
> * save those stats separately
> * run inference using that version
> 
> If the policy behaves correctly in that setup, it would strongly suggest the issue is due to global normalization rather than model behavior. Is validation currently expected to work in this configuration, or is the recommended workflow still to train on the full dataset and evaluate separately? Happy to provide a minimal repro or help test a fix if useful.

Thank you very much for this feedback, it is very relevant and gives me a very serious lead to explore.

**Comment by jashshah999** [comment]
"Robot doesn't move" can be caused by frozen/zero actions in training data that teach the policy to do nothing. [lerobot-doctor](https://github.com/jashshah999/lerobot-doctor) detects this:

```bash
lerobot-doctor /path/to/dataset --checks actions,anomalies
```

Flags: frozen actions (same value >80% of timesteps), stuck actuators, zero-variance action dimensions, and near-duplicate episodes.

https://github.com/jashshah999/lerobot-doctor
